geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/AccumuloIndexAdapter.scala

```diff
@@ -0,0 +1,137 @@
+import org.apache.accumulo.core.data.{Key, Mutation, Range, Value}
+import org.apache.hadoop.io.Text
+import org.geotools.factory.Hints
+import org.locationtech.geomesa.accumulo.data.{AccumuloDataStore, AccumuloFeature}
+import org.locationtech.geomesa.accumulo.index.AccumuloIndexAdapter.ScanConfig
+import org.locationtech.geomesa.accumulo.iterators._
+import org.locationtech.geomesa.index.api.{FilterStrategy, QueryPlan}
+import org.locationtech.geomesa.index.index.IndexAdapter
+import org.locationtech.geomesa.utils.collection.CloseableIterator
+import org.locationtech.geomesa.utils.index.VisibilityLevel
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter.Filter
+
+trait AccumuloIndexAdapter extends IndexAdapter[AccumuloDataStore, AccumuloFeature, Mutation, Range] {
+
+  this: AccumuloFeatureIndex =>
+
+  import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
+
+  override protected def createInsert(row: Array[Byte], feature: AccumuloFeature): Mutation = {
+    val mutation = new Mutation(row)
+    feature.fullValues.foreach(v => mutation.put(v.cf, v.cq, v.vis, v.value))
+    if (hasPrecomputedBins) {
+      feature.binValues.foreach(v => mutation.put(v.cf, v.cq, v.vis, v.value))
+    }
+    mutation
+  }
+
+  override protected def createDelete(row: Array[Byte], feature: AccumuloFeature): Mutation = {
+    val mutation = new Mutation(row)
+    feature.fullValues.foreach(v => mutation.putDelete(v.cf, v.cq, v.vis))
+    if (hasPrecomputedBins) {
+      feature.binValues.foreach(v => mutation.putDelete(v.cf, v.cq, v.vis))
+    }
+    mutation
+  }
+
+  override protected def range(start: Array[Byte], end: Array[Byte]): Range = {
+    // index api defines empty start/end for open-ended range - in accumulo, it's indicated with null
+    val startKey = if (start.length == 0) { null } else { new Key(new Text(start)) }
+    val endKey = if (end.length == 0) { null } else { new Key(new Text(end)) }
+    // index api defines start row inclusive, end row exclusive
+    new Range(startKey, true, endKey, false)
+  }
+
+  override protected def rangeExact(row: Array[Byte]): Range = Range.exact(new Text(row))
+
+  override protected def scanPlan(sft: SimpleFeatureType,
+                                  ds: AccumuloDataStore,
+                                  filter: FilterStrategy[AccumuloDataStore, AccumuloFeature, Mutation],
+                                  hints: Hints,
+                                  ranges: Seq[Range],
+                                  ecql: Option[Filter]): QueryPlan[AccumuloDataStore, AccumuloFeature, Mutation] = {
+    if (ranges.isEmpty) { EmptyPlan(filter) } else {
+      val table = getTableName(sft.getTypeName, ds)
+      val numThreads = queryThreads(ds)
+      val dedupe = hasDuplicates(sft, filter.primary)
+      val ScanConfig(iters, cf, eToF, reduce) = scanConfig(sft, hints, ecql, dedupe)
```

Comment:
Since I was right, it wasn't confusing to me;).  If the class 'ScanConfig' has some docs or nicely readable bits, I'd say it is no big deal.